### ISSUE TITLE
[BUG] WinUI statecontainer crashes app 

### ISSUE THREAD

**Issue by MrDeej** [CommunityToolkit/Maui#806]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

If you click too fast on the statecontainer change state button in sample, the app crashes due to task cancellation beeing thrown to cancel the state changing task

### Expected Behavior

It should just cancel the state change task. Without an exception

### Steps To Reproduce

Open windows
Navigate to statecontainer
Click any state changing button fast enough that a new state is requested before the current state is finished with transitioning.

App hard crash happens in windows, due to WinUI missing a global unhandled exception for async tasks (they have for voids though)

### Link to public reproduction project repository

Use the sample app in this repo

### Environment

```markdown
- .NET MAUI CommunityToolkit:
- OS:
- .NET MAUI:
All environments and version of Maui which creates a WinUI application
```

### Anything else?

I can fix this

**Comment by MrDeej** [comment]
This also hard crash:
![image](https://user-images.githubusercontent.com/16320446/205370753-fbb82ae1-49f3-409f-bbff-4c9e5851de43.png)

**Comment by MrDeej** [comment]
None of these works on windows:
![image](https://user-images.githubusercontent.com/16320446/205370944-fd3f7e08-d5ec-49ca-9f01-7feaffa35419.png)

**Comment by MrDeej** [comment]
here also hard crash:
![image](https://user-images.githubusercontent.com/16320446/205371418-5778756b-0c75-478a-9547-5210a3b34fa3.png)

**Comment by MrDeej** [comment]
Some things looks really odd on Windows:
![image](https://user-images.githubusercontent.com/16320446/205371710-ad28e75d-f67a-4057-b217-39be46ee3d0a.png)
![image](https://user-images.githubusercontent.com/16320446/205371727-9eca8049-4436-4001-948a-fb9cf477f53e.png)

**Comment by MrDeej** [comment]
This is wat happends if you dont catch all exceptions in WinUI:
![image](https://user-images.githubusercontent.com/16320446/205374073-b0719801-a7a3-40d6-93f7-49837634f3a0.png)

All WinUI devs hate this dialog because it gives no insight to what actually threw the exception. So if this happends randomly, you may give up the platform in frustration

**Comment by VladislavAntonyuk** [comment]
As for Snackbar and toast, we use Windows notifications, so make sure you disabled Do not disturb mode

**Comment by VladislavAntonyuk** [comment]
As for statusBar color, you should at least get warning, that platform is not supported.

We prefer throwing exception if something is not supported or behaves incorrectly.

As for Converters, Behaviors and Animations you can suppress the exception or return default value.

As for Controls it is not supported.

**Comment by MrDeej** [comment]
But you really cannot expect to pack this code in a nuget and have a developer implement it, just to get a non-descriptive app crash bug. 

This is unheard of

**Comment by brminnick** [comment]
@MrDeej It sounds like you're upset with how the Visual Studio debugger handles unhandled Exceptions. 

I agree that it is a poor developer experience and they need to improve their tooling. 

We do not control Visual Studio tooling in this CommunityToolkit organization. 

I encourage you to open an issue with the Visual Studio team to improve their rolling. 

Going forward, we will keep this conversation focused on the bug reported in the Issue, not on Visual Studio tooling.

**Comment by MrDeej** [comment]
No, I am upset that there is runtime bugs in this nuget that will crash an app without exception, and it most likely will be random also.

Nothing to do with visual studio. This will behave like I describe even if you installed from the windows store

**Comment by MrDeej** [comment]
Have none of you guy tried to replicate the bug? It takes like 30 seconds to run and crash the app

**Comment by brminnick** [comment]
I'm still looking into the best way to resolve this Issue, but wanted to share a workaround for anyone experiencing this problem in the meantime:

### Workaround

Set `ShouldAnimateOnStateChange` to `false`:

**XAML**
```xml
mct:StateContainer.ShouldAnimateOnStateChange="False"
```
**C#**
```cs
StateContainer.SetShouldAnimateOnStateChange(myLayout, false);
```

### Explanation 

The `TaskCanceledException` is thrown when `CurrentState` changes before the fade animation has completed. 

When `ShouldAnimateOnStateChange == false`, the `SwitchToState()` and `SwitchToContent()` methods run synchronously (eg They don't need to `await` the fade animations). This means the methods always complete immediately.

**Comment by MrDeej** [comment]
If it is missing a view for a string-typed state, it should display an error template in the UI. 

I am referring to the GetViewForState. 

Also, if possible, do not use string for states.

**Comment by brminnick** [comment]
Hi @MrDeej. Are you reporting a separate bug for `StateContainer`? If so, please open a new [Bug Report](https://github.com/CommunityToolkit/Maui/issues/new?assignees=&labels=bug%2Cunverified&template=bug_report_template.yml&title=%5BBUG%5D+) providing a reproduction sample so that we can fix the bug. 

If this is a question about the expected behavior of `StateContainer`, we recommend opening a [New Discussion](https://github.com/communitytoolkit/maui/discussions/new). 

I have answered your questions below to the best of my abilities, inferring your intent. If I misunderstood your intent, please either open a new Bug Report or a Discussion where we can go into more detail. 

> If it is missing a view for a string-typed state, it should display an error template in the UI.
> 
> I am referring to the GetViewForState.

`CommunityToolkit.Maui`purposely does not display errors on the UI; we defer displaying error messages on the UI to the developer consuming `CommunityToolkit.MAUI`. It is the job of the developer consuming this library to gracefully handle the errors we expose via exceptions in a try/catch block and then decide how to handle the thrown exceptions in the manner that best suits their application. 

Sometimes the best way to handle an Exception is to display an Error UI to the user, sometimes it is best handled in business logic. That behavior is not ours to define as it is specific to each application. 

A `StateContainerException` is the expected behavior for when a non-defined state is provided to `StateContainer`.

If the developer implementing `StateContainer` changes `CurrentState` to a state that is not defined in `StateViews`, `GetViewForState` throws a `StateContainerException` because the developer provided `StateContainer` with an invalid state to which we cannot handle:

https://github.com/CommunityToolkit/Maui/blob/0540501a9d3cfdb5bb80a407fbc0e42fbdde6a5d/src/CommunityToolkit.Maui/Layouts/StateContainer/StateContainerController.shared.cs#L153

To avoid this `StateContainerException`, be sure to define all valid states in `StateViews` and be sure not to provide an invalid state. 

> Also, if possible, do not use string for states.

We purposely choose to use `string` as the type for `StateKey` to allow developers the flexibility of defining their own states. 

Other implementations that use an `enum` are limited to the states pre-defined by the `enum` which doesn't allow the developer to define additional states.

**Comment by brminnick** [comment]
@MrDeej We just opened a new Proposal on StateContainer + Exceptions that I'd love for you to take a look at.

It is a breaking change, but I think it'll finally fix the headaches you were having with StateContainer.

Leave a comment and let us know what you think! 

https://github.com/CommunityToolkit/Maui/issues/1027
